Type _sympy/functions.py [1/n]

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136205

Signed-off-by: Bob Ren <bobren@fb.com>

I was chatting with @jamesjwu about strategies to learn the code and he suggested adding types to some files. This stack of PRs adds types to _sympy/functions.py

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

bobrenjc93 committed:
Type _sympy/functions.py [1/n]

Signed-off-by: Bob Ren <bobren@fb.com>

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Type _sympy/functions.py [1/n]"


Signed-off-by: Bob Ren <bobrenfb.com>

I was chatting with jamesjwu about strategies to learn the code and he suggested adding types to some files. This stack of PRs adds types to _sympy/functions.py

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -76,26 +79,29 @@
-            r = sympy.Float(float(r))
+            try:
+                r = sympy.Float(r)
+            except TypeError:
```
What's going on here?

bobrenjc93 replied:
fixed

ezyang commented:
Meta-comment: typing code that interacts with Sympy is often quite difficult, as Sympy is not very static type friendly. Something to be aware of; sometimes the juice is not worth the squeeze.

aorenste reviewed (commented):
I'm not sure if this is meant to completely annotate the types in this file, but if it is you should remove the "mypy: allow-untyped-defs" from the top.

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -162,28 +172,27 @@ class FloorDiv(sympy.Function):
         return self.args[1]
 
-    def _sympystr(self, printer):
+    def _sympystr(self, printer: Any) -> str:
```
should be sympy.printing.printer.Printer I think

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -76,26 +79,29 @@
 
 
-def fuzzy_eq(x, y):
+def fuzzy_eq(x: Any, y: Any) -> Optional[bool]:
```
Any here should be Optional[bool]

ezyang commented:
Blocked on https://github.com/pytorch/pytorch/pull/136205/files#r1763667817

bobrenjc93 committed:
Update on "Type _sympy/functions.py [1/n]"


Signed-off-by: Bob Ren <bobrenfb.com>

I was chatting with jamesjwu about strategies to learn the code and he suggested adding types to some files. This stack of PRs adds types to _sympy/functions.py

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

bobrenjc93 commented:
Sorry @ezyang @aorenste I'm still getting used to the GH workflow and didn't mean to publish this PR. My typical workflow is to send draft PRs and only after CI passes, do I publish it. But GH publishes the PR when you add reviewers (which isn't the case in phabricator).

bobrenjc93 committed:
Update on "Type _sympy/functions.py [1/n]"


Signed-off-by: Bob Ren <bobrenfb.com>

I was chatting with jamesjwu about strategies to learn the code and he suggested adding types to some files. This stack of PRs adds types to _sympy/functions.py

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

aorenste commented:
> Sorry @ezyang @aorenste I'm still getting used to the GH workflow and didn't mean to publish this PR. My typical workflow is to send draft PRs and only after CI passes, do I publish it. But GH publishes the PR when you add reviewers (which isn't the case in phabricator).

You can use the ghstack option `--draft` to publish as a draft.

bobrenjc93 commented:
> > Sorry @ezyang @aorenste I'm still getting used to the GH workflow and didn't mean to publish this PR. My typical workflow is to send draft PRs and only after CI passes, do I publish it. But GH publishes the PR when you add reviewers (which isn't the case in phabricator).
> 
> You can use the ghstack option `--draft` to publish as a draft.

That's what I did! But it seems like adding reviewers automatically promotes the PR from draft status to in review.

Skylion007 commented on torch/utils/_sympy/functions.py:
```diff
@@ -19,6 +30,8 @@
 from .numbers import int_oo
 
 
+T = TypeVar("T", bound=SupportsFloat)
```
Unlikely this TypeVar should be public, should probably begin with `_T` ('in both variable and name)` so it's not exported.

Skylion007 reviewed (approved):
Other than the nit: glad to see the type ignores disappear.

jamesjwu commented on torch/utils/_sympy/functions.py:
```diff
@@ -76,9 +89,9 @@
 
 
-def _keep_float(f):
+def _keep_float(f: Callable[..., T]) -> Callable[..., Union[T, sympy.Float]]:
```
If T supports float already, what does it mean for the return type of the callable to be Union(T, sympy.Float)?

jamesjwu reviewed (approved):
Good to go, but do change T to be private so it's not exported

bobrenjc93 committed:
Update on "Type _sympy/functions.py [1/n]"


Signed-off-by: Bob Ren <bobrenfb.com>

I was chatting with jamesjwu about strategies to learn the code and he suggested adding types to some files. This stack of PRs adds types to _sympy/functions.py

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge

bobrenjc93 committed:
Update on "Type _sympy/functions.py [1/n]"


Signed-off-by: Bob Ren <bobrenfb.com>

I was chatting with jamesjwu about strategies to learn the code and he suggested adding types to some files. This stack of PRs adds types to _sympy/functions.py

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge
